src/common/AndroidUtils.ts

```diff
@@ -236,42 +236,47 @@ export class AndroidSDKUtils {
     > {
         return new Promise<AndroidPackage>(async (resolve, reject) => {
             try {
+                const minSupportedAndroidRuntime: Version = new Version(
+                    androidConfig.minSupportedAndroidRuntime
+                );
                 const packages = await AndroidSDKUtils.fetchInstalledPackages();
                 if (packages.size < 1) {
                     return reject(
                         new Error(
-                            `Could not find android api packages. Requires any one of these [${
-                                androidConfig.supportedRuntimes[0]
-                            } - ${
-                                androidConfig.supportedRuntimes[
-                                    androidConfig.supportedRuntimes.length - 1
-                                ]
-                            }]`
+                            `Could not find any supported Android API packages. Minimum supported Android API package version is ${androidConfig.minSupportedAndroidRuntime}`
                         )
                     );
                 }
 
-                const filteredList: string[] = androidConfig.supportedRuntimes.filter(
-                    (runtimeString) =>
-                        packages.get('platforms;' + runtimeString) !== null
-                );
-                if (filteredList.length < 1) {
+                const matchingKeys: string[] = [];
+                packages.forEach((value, key) => {
+                    if (key.toLowerCase().startsWith('platforms;android-')) {
```

Comment:
This is the wrong place to do this. We should not let the messy parsing details of `sdkmanager --list` leak outside of [`AndroidPackage.parseRawPackagesString()`](https://github.com/forcedotcom/lwc-dev-mobile/blob/21360190c4f6fdcbea505f1a12a3e97498a95e4a/src/common/AndroidTypes.ts#L42-L93). The `AndroidPackage` class is responsible for hosting the normalized parsed data as queryable data fields.

I propose we add two new fields (and their parsing) to the `AndroidPackage` class:

- `packageType`: An enumeration, or well-known strings, representing the "package type" (e.g. "`platforms`") leading value. Incidentally, I think we should only parse for the values we care about. Namely: "`platforms`" and ["`system-images`"](https://github.com/forcedotcom/lwc-dev-mobile/blob/21360190c4f6fdcbea505f1a12a3e97498a95e4a/src/common/AndroidUtils.ts#L665).
- `apiVersion`: The API version associated with this entry (e.g. `android-30`). I'm cool if we either leave this only parsed to `android-30` (rather than `30`) in `parseRawPackagesString()` and parse it further later, or parse it down to `30` in that method. Whichever you think works better.

Reply:
I'm not sure if I'm convinced of doing this change. The intended purpose of **`parseRawPackagesString`** is to parse the raw string and return a parsed data and I think we should keep the logic as simple as that.

We can however change the type of the returned data from **`Map<string, AndroidPackage>`** to something else that would allow manipulating the data more easily. For example we could have something like this:

```
export class AndroidPackages {
    public platforms: AndroidPackage[];
    public systemImages: AndroidPackage[];
    ........
}

export class AndroidPackage {
    public name: string;
    public path: string;
    public version: Version;
    public description: string;
    public location: string;
    ........
}
```

Then update **`parseRawPackagesString`** to return **`AndroidPackages`** instead of **`Map<string, AndroidPackage>`**. Then the caller would do **`packages.filter((item) => item.version.sameOrNewer(minSupportedAndroidRuntime))`**

Reply:
That sounds really good too. That's the heart of what I'd like to see: better data compartmentalization than we have here. I like your suggestion.